io/src/main/java/smile/io/CSV.java

```diff
@@ -100,17 +103,26 @@ public DataFrame read(Path path) throws IOException {
      * @param limit reads a limited number of records.
      */
     public DataFrame read(Path path, int limit) throws IOException {
+        return read(new InputStreamReader(Files.newInputStream(path), charset), limit);
+    }
+
+    /**
+     * Reads a limited number of records from a CSV file.
+     * @param path a CSV file path.
+     * @param limit reads a limited number of records.
+     */
+    public DataFrame read(Reader reader, int limit) throws IOException {
         if (schema == null) {
             // infer the schema from top 1000 rows.
-            schema = inferSchema(path, Math.min(1000, limit));
+            schema = inferSchema(reader, Math.min(1000, limit));
         }
 
         StructField[] fields = schema.fields();
         List<Function<String, Object>> parser = schema.parser();
 
-        try (CSVParser reader = CSVParser.parse(path, charset, format)) {
+        try (CSVParser csvParser = CSVParser.parse(reader, format)) {
```

Comment:
This is troublesome. inferSchema already read some data, which will be missed by CSVParser here. It is why we pass a Path instead of InputStream/Reader.

Reply:
Good call out. I added a call to `reader.reset()` to address. ot all readers are able to be reset, but it should still allow this method to be called with a wider range of inputs